cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -0,0 +1,258 @@
+			}
+			if mig, found := gc.migCache[*instance]; found {
+				return mig, nil
+			}
+			return nil, fmt.Errorf("Instance %+v does not belong to any configured MIG", *instance)
+		}
+	}
+	// Instance doesn't belong to any configured mig.
+	return nil, nil
+}
+
+// RegenerateCacheWithLock triggers cache regeneration under lock.
+func (gc *GceCache) RegenerateCacheWithLock() error {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+
+	return gc.regenerateCache()
+}
+
+// internal method - should only be called after locking on cacheMutex.
+func (gc *GceCache) regenerateCache() error {
+	newMigCache := make(map[GceRef]Mig)
+
+	for _, migInfo := range gc.GetMigs() {
+		mig := migInfo.Config
+		glog.V(4).Infof("Regenerating MIG information for %s", mig.GceRef().String())
+
+		basename, err := gc.GceService.FetchMigBasename(mig.GceRef())
+		if err != nil {
+			return err
+		}
+		gc.updateMigBasename(mig.GceRef(), basename)
+
+		instances, err := gc.GceService.FetchMigInstances(mig.GceRef())
+		if err != nil {
+			glog.V(4).Infof("Failed MIG info request for %s: %v", mig.GceRef().String(), err)
+			return err
+		}
+		for _, ref := range instances {
+			newMigCache[ref] = mig
+		}
+	}
+
+	gc.migCache = newMigCache
+	return nil
+}
+
+// SetResourceLimiter sets resource limiter.
+func (gc *GceCache) SetResourceLimiter(resourceLimiter *cloudprovider.ResourceLimiter) {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+
+	gc.resourceLimiter = resourceLimiter
+}
+
+// GetResourceLimiter returns resource limiter.
+func (gc *GceCache) GetResourceLimiter() (*cloudprovider.ResourceLimiter, error) {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+
+	return gc.resourceLimiter, nil
+}
+
+// GetMachineFromCache retrieves machine type from cache under lock.
+func (gc *GceCache) GetMachineFromCache(machineType string, zone string) *gce.MachineType {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+
+	return gc.machinesCache[MachineTypeKey{zone, machineType}]
+}
+
+// AddMachineToCache adds machine to cache under lock.
+func (gc *GceCache) AddMachineToCache(machineType string, zone string, machine *gce.MachineType) {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+
+	gc.machinesCache[MachineTypeKey{zone, machineType}] = machine
+}
+
+// SetMachinesCache sets the machines cache under lock and returns next recommended refresh time.
+func (gc *GceCache) SetMachinesCache(machinesCache map[MachineTypeKey]*gce.MachineType) time.Time {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+
+	gc.machinesCache = machinesCache
+	gc.machinesCacheLastRefresh = time.Now()
+	return gc.machinesCacheLastRefresh.Add(machinesRefreshInterval)
+}
+
+// Methods that need no lock.
+
+// MachinesCacheFresh return true if the last update to machines cache was within refesh interval.
+func (gc *GceCache) MachinesCacheFresh() bool {
+	return gc.machinesCacheLastRefresh.Add(machinesRefreshInterval).After(time.Now())
```

Comment:
Does this really need no lock? Is resolving a field in struct and calling a method on it atomic in golang? We didn't protect the last refresh timestamp before, because we only ever accessed it from one goroutine anyway. If we're placing it behind a thread-safe interface we may need to reevaluate that.

Reply:
Moved to gceManagerImpl as suggested in other comments